lib/integrations/awsoidc/deploydbservice.go

```diff
@@ -0,0 +1,599 @@
+			Name:       &taskAgentContainerName,
+			LogConfiguration: &ecsTypes.LogConfiguration{
+				LogDriver: ecsTypes.LogDriverAwslogs,
+				Options: map[string]string{
+					"awslogs-group":         "ecs-" + *req.ClusterName,
+					"awslogs-region":        req.Region,
+					"awslogs-create-group":  "true",
+					"awslogs-stream-prefix": *req.ServiceName + "/" + *req.TaskName,
+				},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return taskDefOut.TaskDefinition, nil
+}
+
+// upsertCluster creates the cluster if it doesn't exist.
+// It will update the cluster if it doesn't have the required capacity provider (FARGATE)
+// It will re-create if its status is INACTIVE.
+// If the cluster status is not ACTIVE, an error is returned.
+// The cluster is returned.
+func upsertCluster(ctx context.Context, clt DeployDBServiceClient, clusterName string) (*ecsTypes.Cluster, error) {
+	var cluster *ecsTypes.Cluster
+
+	describeClustersResponse, err := clt.DescribeClusters(ctx, &ecs.DescribeClustersInput{
+		Clusters: []string{clusterName},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inactiveClusterExists := false
+	if len(describeClustersResponse.Clusters) > 0 {
+		cluster = &describeClustersResponse.Clusters[0]
+
+		// If the cluster was recently removed, it might still be listed but with "INACTIVE" status.
+		// Calling CreateCluster activates it.
+		// From AWS Docs:
+		// > INACTIVE The cluster has been deleted. Clusters with an INACTIVE status may remain discoverable in your account for a period of time.
+		if *cluster.Status == clusterStatusInactive {
+			inactiveClusterExists = true
+
+		} else if !slices.Contains(cluster.CapacityProviders, launcTypeFargateString) {
+			// Ensure the required capacity provider (Fargate) is available.
+			putClusterCPResp, err := clt.PutClusterCapacityProviders(ctx, &ecs.PutClusterCapacityProvidersInput{
+				Cluster:           &clusterName,
+				CapacityProviders: requiredCapacityProviders,
+				DefaultCapacityProviderStrategy: []ecsTypes.CapacityProviderStrategyItem{{
+					CapacityProvider: &launcTypeFargateString,
+				}},
+			})
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			cluster = putClusterCPResp.Cluster
+		}
+	}
+
+	// No cluster exists or the cluster is inactive.
+	if len(describeClustersResponse.Clusters) == 0 || inactiveClusterExists {
+		createClusterResp, err := clt.CreateCluster(ctx, &ecs.CreateClusterInput{
+			ClusterName:       &clusterName,
+			CapacityProviders: requiredCapacityProviders,
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		cluster = createClusterResp.Cluster
+	}
+
+	// Anything other than ACTIVE, should throw an error (usually retryable)
+	// Possible status: INACTIVE, PROVISIONING, DEPROVISIONING, FAILED
+	if cluster.Status != nil && *cluster.Status != clusterStatusActive {
+		return nil, trace.Errorf("cluster %q has an invalid status (%s), try again", clusterName, *cluster.Status)
+	}
```

Comment:
Any particular reason why we'd want to error out on `PROVISIONING` state instead of trying to wait on it, at least for a little while?

Reply:
We _can_ add a retry.
However, those clusters are really fast and I've never seen a `PROVISIONING` Status.
So, even if we create and check the status from the creation response, it is already Active

I would rather not add the retry logic to not increase the complexity for little to no gain.

Reply:
I don't have specific knowledge about this service, but I wouldn't rely on optimal or even mediocre perf. If observing a `PROVISIONING` state post-creation is possible, someone is going to hit it. Our users have a proud tradition of thrashing supposedly scalable systems with their workloads.